src/granite/columns.cr

```diff
@@ -0,0 +1,135 @@
+require "json"
+require "uuid"
+
+module Granite::Columns
+  alias SupportedArrayTypes = Array(String) | Array(Int16) | Array(Int32) | Array(Int64) | Array(Float32) | Array(Float64) | Array(Bool)
+  alias Type = DB::Any | SupportedArrayTypes | UUID
+  TIME_FORMAT_REGEX = /\d{4,}-\d{2,}-\d{2,}\s\d{2,}:\d{2,}:\d{2,}/
+
+  module ClassMethods
+    # All fields
+    def fields : Array(String)
+      {% begin %}
+        {% columns = @type.instance_vars.select { |ivar| ivar.annotation(Granite::Column) }.map(&.name.stringify) %}
+        {{columns.empty? ? "[] of String".id : columns}}
+      {% end %}
+    end
+
+    # Columns minus the PK
+    def content_fields : Array(String)
+      {% begin %}
+        {% columns = @type.instance_vars.select { |ivar| (ann = ivar.annotation(Granite::Column)) && !ann[:primary] }.map(&.name.stringify) %}
+        {{columns.empty? ? "[] of String".id : columns}}
+      {% end %}
+    end
+  end
+
+  def content_values
+    parsed_params = [] of Type
+    {% for column in @type.instance_vars.select { |ivar| (ann = ivar.annotation(Granite::Column)) && !ann[:primary] } %}
+      {% ann = column.annotation(Granite::Column) %}
+      parsed_params << {% if ann[:converter] %} {{ann[:converter]}}.to_db {{column.name.id}} {% else %} {{column.name.id}} {% end %}
+    {% end %}
+    parsed_params
+  end
+
+  # Defines a column *decl* with the given *options*.
+  macro column(decl, **options)
+    {% type = decl.type %}
+
+    # Raise an exception if the delc type has more than 2 union types or if it has 2 types without nil
+    # This prevents having a column typed to String | Int32 etc.
+    {% if type.is_a?(Union) && (type.types.size > 2 || (type.types.size == 2 && !type.types.any?(&.resolve.nilable?))) %}
+      {% raise "The type of #{@type.name}##{decl.var} cannot be a Union.  The 'field' macro declares the type as nilable by default.  Use the 'field!' macro to declare a not nilable field." %}
+    {% end %}
+
+    {% column_type = (options[:column_type] && !options[:column_type].nil?) ? options[:column_type] : nil %}
+    {% converter = (options[:converter] && !options[:converter].nil?) ? options[:converter] : nil %}
+    {% primary = (options[:primary] && !options[:primary].nil?) ? options[:primary] : false %}
+    {% auto = (options[:auto] && !options[:auto].nil?) ? options[:auto] : false %}
+    {% auto = (!options || (options && options[:auto] == nil)) && primary %}
+
+    {% nilable = (type.is_a?(Path) ? type.resolve.nilable? : (type.is_a?(Union) ? type.types.any?(&.resolve.nilable?) : (type.is_a?(Generic) ? type.resolve.nilable? : type.nilable?))) %}
+
+    @[Granite::Column(column_type: {{column_type}}, converter: {{converter}}, auto: {{auto}}, primary: {{primary}}, nilable: {{nilable}})]
```

Comment:
@eliasjpr For this initial version I tried to keep the user facing DSL as close to what it is currently.  I'm treating the annotations as an implementation detail for the moment.  If that's something we want to expose more to the end use it is doable since thats pretty much what the `column` macro does.

Just know the macro also does some other stuff for you.  Such as defining the ivars as nilable but defines the getters to be no nilable, and making the setter not be nilable if the column is not nilable.

Reply:
Thanks. I think the team will agree on using `annotations as an implementation detail`.